Implement List component for Notifications

dvdchr opened the pull request:
Resolves #16852
Refs #16740, #16741

## Summary
This PR introduces the new List component to Notifications list, with a feature flag. Here's a summary of the changes:

- `NotificationsViewController` now stores a Boolean property, `usesUnifiedList` which gets assigned at initialization time. This value is refreshed again on every `viewWillAppear`, with a `didSet` property observer that calls `reloadTableViewPreservingSelection()` whenever the value changes.
  - Reason: because `NotificationsViewController` is the root view controller of the notifications tab, we'd need to reload the table manually after the feature flag is toggled. This is why the feature flag value is fetched on `viewWillAppear`.
- All the cell configurations are encapsulated away in `ListTableViewCell+Notifications`. 
- I ended up creating a `ListSimpleOverlayView`, which is a contextless view to be displayed on top of `ListTableViewCell`. This overlay view is contextualized in `ListTableViewCell+Notifications`.
  `ListSimpleOverlayView`, Normal size | `ListSimpleOverlayView`, AX5 size
  --|--
  ![overlay_normal](https://user-images.githubusercontent.com/1299411/126152786-e86dd89a-b0a3-4310-af9f-77d14561a791.png) | ![overlay_ax5](https://user-images.githubusercontent.com/1299411/126152794-7b7cadd6-8b5f-47c7-b01f-bdb31aad17b0.png) 
- In this PR, I added manual bottom separator drawing to `ListTableViewCell`. This is because `NoteTableViewCell` draws its own separator, while `ListTableViewCell` depends on single-lined UITableView separator style. Since this is feature flagged, separator appearance can only be set from the XIB (unless I created a new UITableView instance with the preferred separator style) which is a bit tricky to do programmatically.

## To Test

#### Scenario 1: With enabled feature flag, Notifications list should be displayed with the new design.
- Go to App Settings > Debug and ensure that `Unified Comments and Notifications List` is enabled.
- Go to notifications tab.
- Verify that the list is displayed with the new design.
- Do a trailing swipe, and trash a comment notification (preferably yours, and on a test site).
- Verify that the overlay view appears, and you can undo the action successfully.

#### Scenario 2: With disabled feature flag, Notifications list should be displayed with the old design.
- Go to App Settings > Debug and ensure that `Unified Comments and Notifications List` is disabled.
- Go to notifications tab.
- Verify that the list is displayed with the old design.
- Do a trailing swipe, and trash a notification.
- Verify that the overlay view appears, and you can undo the action successfully.

## Regression Notes
1. Potential unintended areas of impact
Notifications list not showing correctly (with the old component) when the feature flag is off.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested both scenarios (with and without the feature flag)

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Enable unified list flag for local dev and integration builds

dvdchr committed:
Implement List table header in Notifications

dvdchr committed:
Improve ListTableViewCell layout for minimal contents

dvdchr committed:
Implement List table view cell in Notifications (wip undo overlay)

dvdchr committed:
Add manual separator line for backward compatibility with Notifications

dvdchr committed:
Add undo overlay capability for List component

dvdchr committed:
Merge branch 'develop' into issue/16852-notifications-list-component

dvdchr committed:
Add some comments for ListSimpleOverlayView

dvdchr committed:
Rename +Notification to +Notifications for consistency

dvdchr committed:
Remove handling for pending comment indicators in Notifications

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -319,8 +341,11 @@ class NotificationsViewController: UITableViewController, UIViewControllerRestor
+        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
+
+        guard type(of: cell) == expectedType else {
             fatalError()
```
I know this is legacy, but we tend not to `fatalError()` anymore, and instead `return UITableViewCell()`. If you like, you can log a message, something like `DDLogError("Error getting Notification table cell")`.

dvdchr replied:
Noted, will do! 🙂

dvdchr replied:
Done here: https://github.com/wordpress-mobile/WordPress-iOS/pull/16885/commits/c2e5665e291b16065b9edb5d2fba013277428dde

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -1139,12 +1187,27 @@ extension NotificationsViewController: WPTableViewHandlerDelegate {
+
+            guard let noteCell = cell as? NoteTableViewCell else {
                 continue
             }
```
This structure is a bit hard to read/follow. Wouldn't simple `if` statements be clearer?
```
            if usesUnifiedList,
               let listCell = cell as? ListTableViewCell{
                let isLastRow = tableViewHandler.resultsController.isLastIndexPathInSection(indexPath)
                listCell.showsBottomSeparator = !isLastRow
                return
            }
            
            if let noteCell = cell as? NoteTableViewCell {
                let isLastRow = tableViewHandler.resultsController.isLastIndexPathInSection(indexPath)
                noteCell.showsBottomSeparator = !isLastRow
            }
```

dvdchr replied:
Definitely! thanks for that. Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/16885/commits/c2e5665e291b16065b9edb5d2fba013277428dde.

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Views/ListTableViewCell+Notifications.swift:
```diff
@@ -0,0 +1,75 @@
+
+        // snippet text
+        snippetText = notification.renderSnippet()?.string ?? String()
+    }
```
This is so nitpicky, but I think these method comments are unnecessary as it's clear what's going on, making them a bit distracting.

dvdchr replied:
Noted. I just realized myself that the comments here didn't really add anything 😅 It's a habit of mine to write out comments first and add the code later, but forgot to clean it up after. Anyways, this is done in https://github.com/wordpress-mobile/WordPress-iOS/pull/16885/commits/c2e5665e291b16065b9edb5d2fba013277428dde!

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Views/ListTableViewCell+Notifications.swift:
```diff
@@ -0,0 +1,75 @@
+
+    /// Creates a pre-styled overlay view based on `ListSimpleOverlayView`.
+    /// This will be used to show options for the user to revert the action performed on the notification.
+    private func makeUndoOverlayView() -> ListSimpleOverlayView {
```
Just a thought - you have a `private extension`, this could go in there as well, keeping all the `private`s together.

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/16885/commits/c2e5665e291b16065b9edb5d2fba013277428dde 🙏

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift:
```diff
@@ -494,6 +519,13 @@ private extension NotificationsViewController {
 
     func setupTableView() {
+        // Since this view controller is a root view controller for the notifications tab, both `NoteTableViewCell` and the new List components
+        // need to be registered to handle feature flag changes. When the feature is fully rolled out, let's remove NoteTableViewCell.
```
👍

ScoutHarris commented on WordPress/Classes/ViewRelated/Notifications/Style/WPStyleGuide+Notifications.swift:
```diff
@@ -18,6 +18,9 @@ extension WPStyleGuide {
         }
 
+        // ListTableViewCell
+        public static let unreadIndicatorColor = UIColor.muriel(color: .init(name: .blue, shade: .shade30))
```
This can just be `UIColor.primaryLight`.

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/16885/commits/c2e5665e291b16065b9edb5d2fba013277428dde. Thanks!

ScoutHarris commented:
Hi @mattmiklic .

I don't know if you've seen this yet. If you have, my apologies. But I was wondering about the space above the section headers. 

When the table first appears, the first section header feels a bit squished under the filter tab bar.

| First - Old | First - New |
|--------|-------|
| <img width="472" alt="old" src="https://user-images.githubusercontent.com/1816888/126383771-744fb784-e79f-4740-9fa8-e4ccfa1fefde.png"> | <img width="472" alt="new" src="https://user-images.githubusercontent.com/1816888/126383781-0ae3fec6-b09b-45fd-8bb5-52d71c0b2a80.png"> |

And when scrolled, it also feels squished at the top.

| Scrolled - Old | Scrolled - New |
|--------|-------|
| <img width="472" alt="Screen Shot 2021-07-20 at 1 20 39 PM" src="https://user-images.githubusercontent.com/1816888/126383896-542e9fae-1532-4fcf-b991-d674cc362e17.png"> | <img width="474" alt="Screen Shot 2021-07-20 at 1 20 08 PM" src="https://user-images.githubusercontent.com/1816888/126383911-b1929c75-39a5-4e50-b60b-869e76a41e73.png"> |

Just wondering what your thoughts are on this.

(@dvdchr - if Matt requests any changes, please do them separately from this PR.)

mattmiklic commented:
> I don't know if you've seen this yet. If you have, my apologies. But I was wondering about the space above the section headers.
>
> When the table first appears, the first section header feels a bit squished under the filter tab bar.

@ScoutHarris though this matches the designs, I see what you mean. It could help if we added 10pt of space at the top of the list, so the first (and only the first) header has a bit more space above it. The ones that fall later in the list look good.

dvdchr committed:
Merge branch 'develop' into issue/16852-notifications-list-component

dvdchr committed:
Add files to Jetpack target

dvdchr committed:
Add changes based on review

- https://github.com/wordpress-mobile/WordPress-iOS/pull/16885#discussion_r673378616
- https://github.com/wordpress-mobile/WordPress-iOS/pull/16885#discussion_r673385683
- https://github.com/wordpress-mobile/WordPress-iOS/pull/16885#discussion_r673398117
- https://github.com/wordpress-mobile/WordPress-iOS/pull/16885#discussion_r673400957
- https://github.com/wordpress-mobile/WordPress-iOS/pull/16885#discussion_r673408429

dvdchr commented:
> It could help if we added 10pt of space at the top of the list, so the first (and only the first) header has a bit more space above it.

I'll try to make this changes separately later. 🙂

ScoutHarris reviewed (approved):
Thanks for all the changes @dvdchr . This looks great!

:shipit:

dvdchr commented:
Thanks @ScoutHarris !
